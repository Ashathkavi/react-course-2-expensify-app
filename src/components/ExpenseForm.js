import React from 'react'
import moment from 'moment'
import DatePicker from 'react-datepicker'


const now = moment()
//console.log(now.format('MMM do, YYYY'))

class ExpenseForm extends React.Component{
    constructor(props){
        super(props)
        this.state= {
            description:props.expense ? props.expense.description : '', 
            note:props.expense ? props.expense.note : '', 
            amount:props.expense ? (props.expense.amount / 100).toString() : '', 
            createdAt:props.expense ?  moment(props.expense.createdAt).toDate() : moment().toDate(),
            error:''
        }
        
    }
    
    
    onDescriptionChange = (e) => { 
        e.persist()
        this.setState(()=>({description:e.target.value}))
    }
    onNoteChange = (e) => {
        e.persist()
        this.setState(()=>({note:e.target.value}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/gm)){
            this.setState(()=>({amount}))
        }        
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}))
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(()=>({calenderFocused:focused}))
    }
    onSubmit = (e) => {
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please provide description and amount.'}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount, 10) * 100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
        }
    }
    render(){
        //console.log(this.state.description)
        return(
            <div>
            {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit = {this.onSubmit}>
                    <input 
                        typr="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}                        
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}                        

                    />
                    <DatePicker
                        selected = {this.state.createdAt}
                        onChange = {this.onDateChange}
                        closeOnScroll={e => e.target === document}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>

                </form>
            </div>
        )
    }
}

export default ExpenseForm