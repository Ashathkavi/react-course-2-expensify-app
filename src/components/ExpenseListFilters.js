import React from 'react'
import {connect} from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import DatePicker from 'react-datepicker'
import moment from 'moment'

export class ExpenseListFilters extends React.Component{

    onDateChange = (dates) => {
        let [startDate, endDate] = dates;
        startDate = moment(startDate).valueOf()
        endDate =  moment(endDate).valueOf()
        //console.log(startDate,'start',endDate,'end')
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onSortChange = (e) => {
        e.target.value === "date" ? 
            this.props.sortByDate() 
        : 
            this.props.sortByAmount()
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DatePicker
                    selected={this.props.filters.startDate}
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onChange= {this.onDateChange}
                    selectsRange
                    inline
                />

            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return{
        filters:state.filter
    }
} 

const mapDispatchToProps = (dispatch) => ({
    setTextFilter:(text) => dispatch(setTextFilter(text)), 
    sortByAmount:() => dispatch(sortByAmount()), 
    sortByDate:() => dispatch(sortByDate()), 
    setStartDate:(startDate) => dispatch(setStartDate(startDate)), 
    setEndDate:(enDate) => dispatch(setEndDate(enDate))
})

const connectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)

export default connectedExpenseListFilters