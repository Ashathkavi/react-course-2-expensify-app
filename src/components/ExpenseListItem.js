import React from 'react'
import {Link} from 'react-router-dom'


const ExpenseListItem = (props) => (
    <div>
        
        <Link to= {`/edit/${props.id}`}>
            <h4>Description: {props.description}</h4>
        </Link>
        <p>Amount: {props.amount}</p>
        <p>CreatedAt: {props.createdAt}</p>


    </div>
)


export default ExpenseListItem