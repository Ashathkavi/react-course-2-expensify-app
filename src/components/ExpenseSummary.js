import numeral from 'numeral'
import {connect} from 'react-redux'
import React from 'react'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'



export const ExpenseSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const fomattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00')
    return (
        <div>
            <p>Viewing {expenseCount} {expenseWord} totaling {fomattedExpensesTotal}</p>
        </div>
    )
}

const mapStatetoProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filter)
    return{
        expenseCount:visibleExpenses.length,
        expensesTotal:selectExpensesTotal(visibleExpenses)
    }
}

const connectedExpenseSummary  = connect(mapStatetoProps)(ExpenseSummary)
export default connectedExpenseSummary