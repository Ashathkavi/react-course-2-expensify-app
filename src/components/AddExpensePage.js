import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {startAddExpense} from '../actions/expenses'

export class AddExpensePage extends React.Component{
    onSubmit=(expense) => {
        //props.dispatch(startAddExpense(expense))
        this.props.startAddExpense(expense)
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
                This is from AddExpensePage component
                <ExpenseForm onSubmit={this.onSubmit}  />
            </div>
        )
    }
}


// this function makes the AddExpensePage component testable
const mapDispatchToProps = (dispatch) => ({
    startAddExpense:(expense) =>dispatch(startAddExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage)