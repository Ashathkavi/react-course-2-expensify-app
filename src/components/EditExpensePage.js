import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {editExpense, startEditExpenses} from '../actions/expenses'
import {startRemoveExpenses} from '../actions/expenses'

export class EditExpensePage extends React.Component{
    

    onSubmit = (expense) => {
        this.props.editExpense(this.props.match.params.id,expense)
        this.props.history.push('/')
    }

    onRemove = () => {
        this.props.removeExpense({id:this.props.expense.id});
        this.props.history.push('/')
    }

    render(){
        return(
            <div>
                This is from EditExpensePage component
                <button onClick={this.onRemove}>Remove</button>
                <ExpenseForm onSubmit={this.onSubmit} expense ={this.props.expense}/>
            </div>
        )
    }
}


const mapStateToProps = (state, props) => {
    return{
        expense:state.expenses.find((expense)=>{
            //console.log(expense)
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    editExpense:(id,expense) =>dispatch(startEditExpenses(id,expense)),
    removeExpense:(date) => dispatch(startRemoveExpenses(date))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage) 