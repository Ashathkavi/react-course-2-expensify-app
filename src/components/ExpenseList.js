import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'


//exported for the testing purporse
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ):(
                props.expenses.map((expense)=>(
                    <ExpenseListItem key={expense.id} {...expense}/>
                ))
            )
            
        }

    </div>
)


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    }
}


const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList