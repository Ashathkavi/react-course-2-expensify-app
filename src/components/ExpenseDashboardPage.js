import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpencesSummary from './ExpenseSummary'

const ExpenseDashboardPage = () => (
    <div>
        <ExpencesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
)

export default ExpenseDashboardPage