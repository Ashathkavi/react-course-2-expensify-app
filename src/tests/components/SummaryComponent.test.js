import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseSummary} from '../../components/ExpenseSummary'


test('should render SummaryComponent with one expenses', ()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount = {1} expensesTotal={235}/>)
    expect(wrapper).toMatchSnapshot()
})


test('should render SummaryComponent with multiple expenses', ()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount = {23} expensesTotal={123214645}/>)
    expect(wrapper).toMatchSnapshot()
})

