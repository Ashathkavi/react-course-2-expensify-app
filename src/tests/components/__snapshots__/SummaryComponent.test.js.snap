// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SummaryComponent with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    23
     
    expenses
     totaling 
    $1,232,146.45
  </p>
</div>
`;

exports[`should render SummaryComponent with one expenses 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totaling 
    $2.35
  </p>
</div>
`;
